frame/bridge/s2s/issuing/src/lib.rs

```diff
@@ -0,0 +1,374 @@
+		type WeightInfo: WeightInfo;
+		type RingCurrency: LockableCurrency<Self::AccountId, Moment = Self::BlockNumber>;
+
+		type ReceiverAccountId: From<[u8; 32]> + Into<Self::AccountId> + Clone;
+		type BridgedAccountIdConverter: Convert<H256, Self::AccountId>;
+		type BridgedChainId: Get<ChainId>;
+		type ToEthAddressT: ToEthAddress<Self::AccountId>;
+		type OutboundPayload: Parameter + Size;
+		type CallEncoder: EncodeCall<Self::AccountId, Self::OutboundPayload>;
+		type FeeAccount: Get<Option<Self::AccountId>>;
+		type MessageSender: RelayMessageCaller<Self::OutboundPayload, RingBalance<Self>>;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Handle dispatch call from dispatch precompile contract
+		///
+		/// When user burn their tokens, this handler will receive the event from dispatch
+		/// precompile contract, and relay this event to the target chain to unlock asset.
+		// TODO: update the weight
+		#[pallet::weight(0)]
+		#[frame_support::transactional]
+		pub fn dispatch_handle(origin: OriginFor<T>, input: Vec<u8>) -> DispatchResultWithPostInfo {
+			let caller = ensure_signed(origin)?;
+
+			// Ensure the input data is long enough
+			ensure!(input.len() >= 8, <Error<T>>::InvalidInput);
+			// Ensure that the user is mapping token factory contract
+			let factory = MappingFactoryAddress::<T>::get();
+			let factory_id = <T as darwinia_evm::Config>::AddressMapping::into_account_id(factory);
+			ensure!(caller == factory_id, <Error<T>>::NotFactoryContract);
+
+			let burn_action = &sha3::Keccak256::digest(&BURN_ACTION)[0..4];
+			if &input[4..8] == burn_action {
+				let burn_info =
+					TokenBurnInfo::decode(&input[8..]).map_err(|_| Error::<T>::InvalidDecoding)?;
+				// Ensure the recipient is valid
+				ensure!(
+					burn_info.recipient.len() == 32,
+					<Error<T>>::InvalidAddressLen
+				);
+
+				let fee = Self::transform_dvm_balance(burn_info.fee);
+				if let Some(fee_account) = T::FeeAccount::get() {
+					// Since fee account will represent use to make a cross chain call, give fee to fee account here.
+					// the fee transfer path
+					// user -> mapping_token_factory(caller) -> fee_account -> fee_fund -> relayers
+					<T as Config>::RingCurrency::transfer(&caller, &fee_account, fee, KeepAlive)?;
+				}
+
+				Self::burn_and_remote_unlock(fee, burn_info)?;
+			} else {
+				log::trace!("No action match this input selector");
+			}
+			Ok(().into())
+		}
+
+		/// Handle remote register relay message
+		/// Before the token transfer, token should be created first
+		#[pallet::weight(0)]
+		pub fn remote_register(origin: OriginFor<T>, token: Token) -> DispatchResultWithPostInfo {
+			let user = ensure_signed(origin)?;
+			let backing = Self::verify_origin(&user)?;
```

Comment:
backing or SourceRoot? Do we need to verify it is SourceRoot here?

Reply:
It will clear if we change the name to `ensure_source_root `.

```suggestion
			Self::ensure_source_root(&user)?;
```

Reply:
It will verify the SourceRoot identity and return an ethereum address called backing address derived from the SourceRoot address and the remote chainid. If the same source ERC20 token address comes from different remote chains, we use the backing address to generate different mapping token addresses.

Reply:
OK